securedrop_client/api_jobs/uploads.py

```diff
@@ -72,34 +71,44 @@ def call_api(self, api_client: API, session: Session) -> str:
             message = "Failed to send reply for source {id} due to Exception: {error}".format(
                 id=self.source_uuid, error=e)
             raise SendReplyJobTimeoutError(message, self.reply_uuid)
+        except SQLAlchemyError as e:
+            # If the draft reply has already made it to the server but we didn't get a 201 response
+            # back, then a reply with self.reply_uuid will exist in the replies table. Otherwise
+            # the reply was never sent and no longer exists in the drafts table (something went
+            # terribly wrong).
+            try:
+                reply_db_object = session.query(Reply).filter_by(uuid=self.reply_uuid).one()
+                logger.debug('''
+                    Draft reply {uuid} has already been sent successfully, skipping: {e}
+                '''.format(uuid=self.reply_uuid, e=e))
+                return reply_db_object.uuid
+            except SQLAlchemyError:
+                message = '''
+                    Failed to send reply {uuid} for source {id} due to Exception: {error}
+                '''.format(id=self.source_uuid, error=e)
+                raise SendReplyJobError(message, self.reply_uuid)
         except Exception as e:
-            message = "Failed to send reply for source {id} due to Exception: {error}".format(
-                id=self.source_uuid, error=e)
+            message = '''
+                Failed to send reply {uuid} for source {id} due to Exception: {error}
+            '''.format(id=self.source_uuid, error=e)
             self._set_status_to_failed(session)
```

Comment:
heads up if an error occurs during the execution of `_set_status_to_failed`, it will be unhandled and will crash the application (we should at least add back the generic exception handling inside the `_set_status_to_failed` method, see #827)

Reply:
i changed it so that `_set_status_to_failed` will only be called if the draft reply exists

Reply:
(still working on cleaning up this code so that people can reason about it more... i'll ping you when ready... i could move the exception handling back but i thought it was easier to read the exception handling by keeping it all in one place. lemme see what i can do to make this easy to reason about, especially wrt exception handling.)

Reply:
but any unexpected exception in this method will now crash the app, no?

Reply:
so my thinking was that by making sure the DraftReply exists like so: 

```py
            # If the draft does not exist because it was deleted locally then do not send the
            # message to the source.
            draft_reply_db_object = session.query(DraftReply).filter_by(
                uuid=self.reply_uuid).one_or_none()
            if not draft_reply_db_object:
                raise Exception('Draft reply {} does not exist'.format(self.reply_uuid))
```

we would not have to make sure it exists again if an unexpected exception is raised before setting it to FAILED, like so:

```py
    def _set_status_to_failed(self, session: Session) -> None:
        try:  # If draft exists, we set it to failed.
            draft_reply_db_object = session.query(DraftReply).filter_by(uuid=self.reply_uuid).one()
            reply_status = session.query(ReplySendStatus).filter_by(
                name=ReplySendStatusCodes.FAILED.value).one()
            draft_reply_db_object.send_status_id = reply_status.id
            session.add(draft_reply_db_object)
            session.commit()
        except SQLAlchemyError as e:
            logger.info('SQL error when setting reply {uuid} as failed, skipping: {e}'.format(
                uuid=self.reply_uuid, e=e))
        except Exception as e:
            logger.error('Unknown error when setting reply {uuid} as failed, skipping: {e}'.format(
                uuid=self.reply_uuid, e=e))
```

But this is a refactor that I removed from this PR (it's back to your original implementation) because I don't want it to hold up the fix.